2000-12-06  Jakub Jelinek  <jakub@redhat.com>

	* optimize.c (copy_body_r): Set STMT_IS_FULL_EXPR_P on EXPR_STMT.

	* g++.old-deja/g++.other/cleanup4.C: New test.

--- gcc/cp/optimize.c.jj	Mon Oct 23 15:24:56 2000
+++ gcc/cp/optimize.c	Wed Dec  6 21:24:27 2000
@@ -289,6 +289,7 @@ copy_body_r (tp, walk_subtrees, data)
 	{
 	  *tp = build_stmt (EXPR_STMT, 
 			    RETURN_EXPR (return_stmt));
+	  STMT_IS_FULL_EXPR_P (*tp) = 1;
 	  /* And then jump to the end of the function.  */
 	  TREE_CHAIN (*tp) = goto_stmt;
 	}
--- gcc/testsuite/g++.old-deja/g++.other/cleanup4.C.jj	Wed Dec  6 15:25:11 2000
+++ gcc/testsuite/g++.old-deja/g++.other/cleanup4.C	Wed Dec  6 15:26:12 2000
@@ -0,0 +1,34 @@
+// Build don't link:
+// Origin: Jakub Jelinek  <jakub@redhat.com>
+// Special g++ Options: -O2 -fno-exceptions
+
+class foo
+{
+public:
+  foo ();
+  ~foo ();
+};
+
+class bar
+{
+public:
+  bar ();
+  bar (const foo&);
+};
+
+int i;
+foo f ();
+
+inline bar x ()
+{
+  switch (i)
+    {
+    case 0: return bar (f ());
+    default: return bar ();
+    }
+}
+
+bar y ()
+{
+  return x ();
+}
